[Week2] ToDo REST API 만들기

Miniminis opened the pull request:
## 과제 목표

* Spring Web 이용해서 ToDo REST API를 만들기
* Marko를 이용해서 웹을 만들고, 만든 REST API와 연결하여 동작하는 ToDo앱을 만들기

### REST API

- ToDo 목록 얻기 - `GET /tasks`
- ToDo 상세 조회하기 - `GET /tasks/{id}`
- ToDo 생성하기 - `POST /tasks`
- ToDo 제목 수정하기 - `PUT/PATCH /tasks/{id}`
- ToDo 삭제하기 - `DELETE /tasks/{id}`

### 프론트엔드

- 할 일 조회하기
- 할 일 추가하기
- 할 일 수정하기
- 할 일 완료하기


## 요구 사항

- [ ] 모든 API 테스트를 통과해야 합니다.
- [ ] 모든 E2E 테스트를 통과해야 합니다.

## 작업 목록

- [ ] java proejct -> spring project 로 변경
- [x] 전체 구조 설계
- [x] CRUD API 제작
- [ ] refactoring
- [x] ~~marko.js 통해서 front 작업~~ -> 이미 작업된 상태로 시작

## 과제 1에서 완료하지 못한 작업 목록
- [x] 테스트 매소드 한글이름 쓰지 않고 `@DisplayName` 사용하기
- [x] 테스트 여러가지 케이스 추가하기
- [x] @Nested 이용하여 계층형 테스트 코드 작성하기 
- [x] POST, PUT, PATCH 시에 Task 가 Null 일 경우 방어하기
- [ ] Task 를 저장하는 자료구조가 List 적절한지 고민해보기 (다른 자료구조를 쓴다면 어떤 점이 좋을까)

## 과제 2에서 피드백을 통해 새롭게 추가된 작업 목록
- [x] PUT 과 PATCH 두 매서드를 한 함수에 매핑할 수 있는지 고민해보기
- [ ] synchronized 키워드에 대해 조사하고 적용할 수 있는 부분에 적용해보기 : newTaskId
- [ ] DTO 를 쓴다면 역할을 명확하게 설명하기
- [ ] 서비스, 컨트롤러, 리포지토리처럼 중요한 역할을 하는 클래스에는 모두 JavaDoc 주석을 작성해보기
- [ ] 나의 코드를 남에게 설명하기 -> comment 이용
- [ ] build.gradle 에 의존성 추가될 때 주석 등을 통해 라이브러리의 용도와 출처 등을 명확히 표기한다.  
- [ ] TaskService 인터페이스로 의존성 역전시키기

## 공부 목록
- [ ] 아샬님 올려주신 강의 듣고 정리하기
  - [Static 대신 객체 사용하기 #1](https://www.youtube.com/watch?v=qRZL2NkXRdo)
  - [Static 대신 객체 사용하기 #2](https://www.youtube.com/watch?v=nZw-1UNfDJQ)
- [ ] [병성님의 javadoc 글 읽고 정리하기](https://resonant-packet-5dd.notion.site/JavaDoc-7b59f9eed97941c89db965cf907c47bb)
- [ ] [세기님의 synchronized 글 읽고 정리하기](https://shineoov.notion.site/synchronized-Java-2ae334a82cb5425f98f60db45431ee5b)
- [ ] [세기님의 spring-developers-tool 읽고 정리하기](https://shineoov.notion.site/Developer-tools-Spring-900c5e9cdec64562bb7a5ed5bd41f25d)
- [ ] [synchronized method 로 taskId ++ 하는 방식과 AtomicLong 과의 차이점](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicLong.html)

Miniminis committed:
test 테스트 코드 실행

Miniminis committed:
feat 프로젝트 전체 구조 설계

Miniminis committed:
doc README 체크목록 추가

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,52 @@
+        return taskService.addTask();
+    }
+
+    @PutMapping("/tasks/{id}")
+    public Task editTask() {
+        return taskService.editTask();
+    }
+
+    @PatchMapping("/tasks/{id}")
+    public Task editTaskElement() {
+        return taskService.editTask();
+    }
```
하나의 함수의 두 가지 Method를 매핑할 수 있을까요?

Miniminis replied:
윤석님이 주신 힌트로 @RequestMapping 을 이용해서 해결했습니다 :) 

```    
@RequestMapping(value = "/tasks/{taskId}", method = {RequestMethod.PUT, RequestMethod.PATCH})
public BaseResponse<Task> editTask(@PathVariable Long taskId, @RequestBody Task task) {
```

hannut91 reviewed (commented):
이번주도 잘 부탁드립니다 ㅎㅎ

hannut91 commented:
진행하는데 어려움이 있으신가요?

Miniminis commented:
> 진행하는데 어려움이 있으신가요?

어제 본업 일정이 바빠서 과제진행을 못했습니다...! 오늘부터 다시 열심히 진행하겠습니다 :)

Miniminis committed:
feat CRUD API 완성

- 과제 1에서 구현했던 코드 그대로 CRUD API 완성

Miniminis committed:
doc README 에 작업 목록 추가 및 보완

Miniminis committed:
feat spring developer tools 의존성 추가

- 앱의 신속한 빌드 & 재시작을 위해서 spring developer tools 의존성을 추가함

Miniminis committed:
refactor PUT, PATCH 두 매소드 하나의 함수에 매핑

Miniminis committed:
test 각 API 별 테스트 코드 추가

- 아직 100% 통과하지 않은 테스트 코드를 추가함

Miniminis committed:
test @Nest 이용하여 계층형 테스트 코드 작성

- 테스트 코드 작성하면서 Task 객체의 정상적인 비교를 위해 equals, hasCode 를 오버라이딩
- 오류케이스의 경우 HttpCode 를 리턴하는 형식에서 오류를 던지는 방식으로 수정
- 예외 메시지 관리를 위해 exception message enum 생성

Miniminis commented:
## 질문

윤석님, 오늘도 고생 많으십니다!
이번 과제 제출에서 @Nest 를 이용하여 계층형 테스트 코드를 작성했는데요, 테스트 코드를 작성하다보니까 궁금한 점이 생겨서 질문 남깁니다! 

### 1. 계층형 테스트코드의 사용성 
음, 확실히 계층형 테스트 코드를 작성하면서 논리적으로 흐름도 정리되고, 중복되는 코드들도 많이 줄어든 것 같은데요. 자바언어의 특성상 계속 중괄호({}) 내부로 계층이 만들어지면서 케이스가 많아질 경우 가독성이 좋지 않을 것 같다고 느꼈습니다. 실무에서 계층형 테스트코드는 자주 사용되는 패턴인가요? 

### 2. 테스트 코드 작성시 한글이름 매소드 사용을 권장하지 않는 이유
지난 과제에서 종립님으로부터 받은 피드백 중 하나가 테스트 매소드 이름에 한글을 쓰지 않고 @DisplayName() 을 이용해보는 것이었는데요. 이번에 해당 내용을 적용해서 과제를 진행하다보니까 @DisplayName("") 내부에 들어갈 한글 이름도 고민해야하고, 매소드 영문 이름도 고민하면서 작성해야해서 오히려 더 고민에 대한 포인트가 늘어났다고 생각되었습니다. 

혹시 이러한 불편함은 저 개인만 느끼는 것일까요 ㅠㅠ? 
실무에서 한글이름 매소드 사용을 권장하지 않는 이유가 있을까요? 

### 3. 테스트 데이터의 초기화
과제 1때와 마찬가지로 저는 할일을 담당하고 있는 Task 클래스와 할일목록을 담당하고 있는 TaskList 클래스를 추가로 만들어 관리를 하고 있었는데요. 할일 목록의 경우, 무분별하게 객체가 생성되어서는 안된다고 생각해서 싱글톤패턴으로 작성을 했습니다.  그런데 테스트 코드를 작성하다보니, 해당 객체가 초기화되지 않아서 다른 테스트 케이스의 경우와 데이터가 섞여버리게 되었습니다 ㅠㅠ 
그렇게 되니 개별 테스트에서는 성공했지만, 전체 클래스 단위에서 테스트를 돌릴 경우 실패해버리게 되었습니다. 
이런 경우, 어떻게 테스트 데이터를 초기화하며 테스트를 할 수 있을까요?

Miniminis committed:
doc README 에 추가 작업목록 및 공부목록 추가

hannut91 commented:
1. 계층형 테스트코드의 사용성

팀의 컨벤션에 맞추는게 더 중요해보이는데요. 저는 많이 사용합니다. 사실 거의 무조건 사용하긴 해요. 왜냐하면 테스트를 작성할 떄 여러가지 상황들을 고려해보기 쉬운 구조라서 그런 것 같아요.

2. 테스트 코드 작성시 한글이름 매소드 사용을 권장하지 않는 이유

테스트를 작성할 때 올바르게 테스트하는 것도 중요하지만, 테스트가 실패했을 때 빨리 아는 것도 매우 중요합니다. 이에 대해서 `테스트 주도 개발로 배우는 객체 지향 설계와 실천`이라는 책에서는 이렇게 얘기하고 있죠.

> 실패하는 테스트가 명확하게 무엇이 실패했고 왜 실패했는지 설명해주면 신속하게 코드를 진단해 오류를 수정할 수 있다. - 308p

그래서 `@DisplayName`을 사용해서 결과를 더 명확하게 표현하기 위해서 입니다. 제가 설명드리는 것 보다 다음을 읽어보시면 좋을 것 같아요

https://johngrib.github.io/wiki/junit5-nested/

한글 이름을 사실 사용해도 아무런 문제가 없는데, 아직 이전 습관에 대한 관성이 있는 것 같아요.

3. 테스트 데이터의 초기화

모든 테스트를 시작하기 전에 테스트 데이터를 초기화 하면 될 것 같아요. 배열의 모든 요소를 삭제하는 것은 어떨까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,84 @@
+
+    @GetMapping("/tasks/{taskId}")
+    public BaseResponse<Task> readTask(@PathVariable Long taskId) {
+        if (isInValidTaskId(taskId)) {
```
`isInvalidTaskId`라고 이름을 수정해야 할 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        if (!newTask.hasValidContent()) {
+            throw new RuntimeException(ExceptionMessages.INVALID_TASK_CONTENT.getMessage());
```
예외를 모두 RuntimeException으로 처리해주고 계시군요. 커스텀 에러 타입을 만들어 보는 것은 어떨까요? 그러면 무엇이 더 좋을까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/domains/Task.java:
```diff
@@ -0,0 +1,71 @@
+        return title;
+    }
+
+    @Override
+    public String toString() {
+        return "Task{" +
+                "id=" + id +
+                ", title='" + title + '\'' +
+                '}';
+    }
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

```java
return "{ id = $id, title = $title }"
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/domains/Task.java:
```diff
@@ -0,0 +1,71 @@
+        return hasValidContent() && Objects.equals(this.title, title);
+    }
+
+    public Task editTaskTitle(String title) {
+        this.title = title;
+        return this;
+    }
```
뭔가 자신을 수정하고, 다시 자신을 반환하는 것을 보니 불변객체로 만들려고 하신 줄 알았는데 아니군요. 어떤 이유 때문에 이렇게 구현하셨나요?

hannut91 commented:
> 다음주에는 이런저런 핑계없이, 깔끔하게 과제를 마무리해서 이번주보다는 훨씬 성장한 사람이 되어있기를 바래본다.

좋네요 ㅎㅎ 다시 시작하면 됩니다! 이번 한 주도 고생 많으셨습니다!
